archive/imo/imo1962_q1.lean

```diff
@@ -0,0 +1,196 @@
+    linarith },
+  { rw ← h1.right,
+    simp [of_digits_append, of_digits_digits],
+    ring },
+  exact nat.pos_of_ne_zero h3
+end
+
+/-
+Now we can solve for each possibility of (nat.digits 10 c).length until we get to the one that works.
+-/
+
+lemma case_0_digit (c n : ℕ) (h1 : (nat.digits 10 c).length = 0)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : false :=
+begin
+  have h3 : 6 * 10 ^ 0 + c = 6 * 10 ^ (nat.digits 10 c).length + c, by rw h1,
+  have h4 : 6 * 10 ^ 0 + c = 4 * (10 * c + 6), by rw [h3, h2.right, h2.left],
+  linarith,
+end
+
+lemma case_1_digit (c n : ℕ) (h1 : (nat.digits 10 c).length = 1)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : false :=
+begin
+  have h3 : 6 * 10 ^ 1 + c = 6 * 10 ^ (nat.digits 10 c).length + c, by rw h1,
+  have h4 : 6 * 10 ^ 1 + c = 4 * (10 * c + 6), by rw [h3, h2.right, h2.left],
+  have h6 : c > 0, by linarith,
+  linarith,
+end
+
+lemma case_2_digit (c n : ℕ) (h1 : (nat.digits 10 c).length = 2)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : false :=
+begin
+  have h3 : 6 * 10 ^ 2 + c = 6 * 10 ^ (nat.digits 10 c).length + c, by rw h1,
+  have h4 : 6 * 10 ^ 2 + c = 4 * (10 * c + 6), by rw [h3, h2.right, h2.left],
+  have h5 : c > 14, by linarith,
+  linarith
+end
+
+lemma case_3_digit (c n : ℕ) (h1 : (nat.digits 10 c).length = 3)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : false :=
+begin
+  have h3 : 6 * 10 ^ 3 + c = 6 * 10 ^ (nat.digits 10 c).length + c, by rw h1,
+  have h4 : 6 * 10 ^ 3 + c = 4 * (10 * c + 6), by rw [h3, h2.right, h2.left],
+  have h5 : c > 153, by linarith,
+  linarith
+end
+
+lemma case_4_digit (c n : ℕ) (h1 : (nat.digits 10 c).length = 4)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : false :=
+begin
+  have h3 : 6 * 10 ^ 4 + c = 6 * 10 ^ (nat.digits 10 c).length + c, by rw h1,
+  have h4 : 6 * 10 ^ 4 + c = 4 * (10 * c + 6), by rw [h3, h2.right, h2.left],
+  have h5 : c > 1537, by linarith,
+  linarith
+end
+
+/- Putting this inline causes a deep recursion error, so we separate it out. -/
+lemma helper_5_digit (c : ℤ) (h : 6 * 10 ^ 5 + c = 4 * (10 * c + 6)) : c = 15384 := by linarith
+
+lemma case_5_digit (c n : ℕ) (h1 : (nat.digits 10 c).length = 5)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : c = 15384 :=
+begin
+  have h3 : 6 * 10 ^ 5 + c = 6 * 10 ^ (nat.digits 10 c).length + c, by rw h1,
+  have h4 : 6 * 10 ^ 5 + c = 4 * (10 * c + 6), by rw [h3, h2.right, h2.left],
+  zify at *,
+  exact helper_5_digit ↑c h4
+end
+
+/-
+linarith fails on numbers this large, so this lemma spells out some of the arithmetic
+that normally would be automated.
+-/
+lemma case_more_digits (c n : ℕ) (h1 : (nat.digits 10 c).length ≥ 6)
+(h2 : n = 10 * c + 6 ∧ 6 * 10 ^ (nat.digits 10 c).length + c = 4 * n) : n ≥ 153846 :=
+begin
+   have h3 : c ≠ 0,
+  { intro h4,
+    have h5 : (nat.digits 10 0).length = 0, by norm_num,
```

Comment:
This should be a lemma in `src/`.

Reply:
also, `simp` can prove this.

Reply:
I removed this line by using `simp` in the following line